packages/gatsby-source-graphql/src/batching/merge-queries.ts

```diff
@@ -0,0 +1,410 @@
+  })
+
+  const mergedQueryDefinition: OperationDefinitionNode = {
+    kind: Kind.OPERATION_DEFINITION,
+    operation: `query`,
+    variableDefinitions: mergedVariableDefinitions,
+    selectionSet: {
+      kind: Kind.SELECTION_SET,
+      selections: mergedSelections,
+    },
+  }
+
+  return {
+    query: {
+      kind: Kind.DOCUMENT,
+      definitions: [mergedQueryDefinition, ...mergedFragmentMap.values()],
+    },
+    variables: mergedVariables,
+  }
+}
+
+/**
+ * Split and transform result of the query produced by the `merge` function
+ */
+export function resolveResult(mergedQueryResult: IQueryResult): IQueryResult[] {
+  const data = mergedQueryResult.data
+
+  return Object.keys(data).reduce(
+    (acc: IQueryResult[], prefixedKey: string): IQueryResult[] => {
+      const { index, originalKey } = Prefix.parseKey(prefixedKey)
+      if (!acc[index]) acc[index] = { data: {} }
+      acc[index].data[originalKey] = data[prefixedKey]
+      return acc
+    },
+    []
+  )
+}
+
+const Visitors = {
+  detectFragmentsWithVariables: (
+    fragmentsWithVariables: Set<string>
+  ): Visitor<ASTKindToNode> => {
+    let currentFragmentName
+    return {
+      [Kind.FRAGMENT_DEFINITION]: {
+        enter: (def: FragmentDefinitionNode): void => {
+          currentFragmentName = def.name.value
+        },
+        leave: (): void => {
+          currentFragmentName = null
+        },
+      },
+      [Kind.VARIABLE]: (): void => {
+        if (currentFragmentName) {
+          fragmentsWithVariables.add(currentFragmentName)
+        }
+      },
+    }
+  },
+  prefixVariables: (prefix: string): Visitor<ASTKindToNode> => {
+    return {
+      [Kind.VARIABLE]: (variable: VariableNode): VariableNode =>
+        prefixNodeName(variable, prefix),
+    }
+  },
+  prefixFragmentNames: (
+    prefix: string,
+    fragmentNames: Set<string>
+  ): Visitor<ASTKindToNode> => {
+    return {
+      [Kind.FRAGMENT_DEFINITION]: (
+        def: FragmentDefinitionNode
+      ): FragmentDefinitionNode | void =>
+        fragmentNames.has(def.name.value) ? prefixNodeName(def, prefix) : def,
+
+      [Kind.FRAGMENT_SPREAD]: (def: FragmentSpreadNode): FragmentSpreadNode =>
+        fragmentNames.has(def.name.value) ? prefixNodeName(def, prefix) : def,
+    }
+  },
+}
+
+function prefixQueryParts(prefix: string, query: IQuery): IQuery {
+  let document: DocumentNode = aliasTopLevelFields(prefix, query.query)
+  const variableNames = Object.keys(query.variables)
+
+  if (variableNames.length === 0) {
+    return { ...query, query: document }
+  }
+
+  const fragmentsWithVariables: Set<string> = new Set()
+
+  document = visit(
+    document,
+    visitInParallel([
+      // Note: the sequence is important due to how visitInParallel deals with node edits
+      Visitors.detectFragmentsWithVariables(fragmentsWithVariables),
+      Visitors.prefixVariables(prefix),
+    ])
+  )
+
+  if (fragmentsWithVariables.size > 0) {
```

Comment:
Is prefixing fragments very expensive? Maybe we should prefix all fragments by default instead of doing two passes?

Reply:
I've been thinking about this. But I suspect that fragments with variables will be **very** rare in the real world.

At the same time, some fragments can be huge (I've seen sites like that) which means that servers will spend more time parsing and validating them.